homeassistant/components/otp/config_flow.py

```diff
@@ -0,0 +1,101 @@
+"""Config flow for One-Time Password (OTP) integration."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import pyotp
+import voluptuous as vol
+
+from homeassistant.auth.mfa_modules.totp import _generate_qr_code
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_CODE, CONF_NAME, CONF_TOKEN
+
+from .const import DEFAULT_NAME, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_TOKEN): str,
+        vol.Required(CONF_NAME, default=DEFAULT_NAME): str,
+    }
+)
+
+STEP_CONFIRM_DATA_SCHEMA = vol.Schema({vol.Required(CONF_CODE): str})
+
+
+class TOTPConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for One-Time Password (OTP)."""
+
+    VERSION = 1
+    user_input: dict[str, Any]
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            # generate new token if empty
+            if not user_input.get(CONF_TOKEN):
+                user_input[CONF_TOKEN] = await self.hass.async_add_executor_job(
+                    pyotp.random_base32
+                )
+
+            await self.async_set_unique_id(user_input[CONF_TOKEN])
+            self._abort_if_unique_id_configured()
+
+            self.user_input = user_input
+            return await self.async_step_confirm()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=self.add_suggested_values_to_schema(
+                data_schema=STEP_USER_DATA_SCHEMA, suggested_values=user_input
+            ),
+            errors=errors,
+            last_step=False,
+        )
+
+    async def async_step_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the confirmation step."""
+
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            try:
+                if await self.hass.async_add_executor_job(
+                    pyotp.TOTP(self.user_input[CONF_TOKEN]).verify, user_input["code"]
+                ):
+                    return self.async_create_entry(
+                        title=self.user_input[CONF_NAME],
+                        data={CONF_TOKEN: self.user_input[CONF_TOKEN]},
+                    )
+
+                errors["base"] = "invalid_code"
+
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        url = await self.hass.async_add_executor_job(
+            pyotp.TOTP(self.user_input[CONF_TOKEN]).provisioning_uri,
+            self.user_input[CONF_NAME],
+            "Home Assistant",
+        )
+        qr_code = await self.hass.async_add_executor_job(_generate_qr_code, url)
```

Comment:
Why do we generate a QR and ask the user to add it to the authenticator app?
Normally the website where you are activating asks to insert a generated code to verify the setup

Reply:
It is to verify the OTP generation is working as expected. If the user somehow typed in the wrong code they would not know because OTP would still generate codes, any string is a valid token to generate OTPs, so this is the only way to validate that the token is correct.

Reply:
removed OTP verification

Reply:
Seems i was wrong about this, it seems to be possible to insert an invalid code that raises en error. Added some basic validation but leaving out the verification with an OTP code